tests/Unit/NumericalAlgorithms/Interpolation/Test_MultiLinearSpanInterpolation.cpp

```diff
@@ -0,0 +1,146 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "Framework/TestHelpers.hpp"
+#include "NumericalAlgorithms/Interpolation/MultiLinearSpanInterpolation.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+template <size_t dim, size_t num_var>
+void test() {
+  //  CAPTURE(Degree);
+
+  // pseudo random number generator
+  std::random_device rd;   // seed
+  std::mt19937 gen(rd());  // generator
+
+  std::uniform_real_distribution<> dist_func(-1., 1.);
+  std::uniform_int_distribution<size_t> dist_int(10, 20);
+
+  // Main idea: The table performs a multi-linear interpolation
+  // Each element in the table will be multi-linear function,
+  // so we can use this directly as a test.
+  // For simplicity, we will use the table to globally reproduce a
+  // single multi-linear function.
+
+
+  std::array< std::array<double, dim>, num_var> A;
+  std::array< double, num_var> b;
+
+  for (size_t i = 0; i < num_var; ++i) {
+    b[i] = dist_func(gen);
+    for (size_t j = 0; j < dim; ++j)
+      A[i][j] = dist_func(gen);
+  }
+
+  auto f = [&](std::array<double, dim> const& x) {
+    std::array<double, num_var> res{0};
+
+    for (size_t i = 0; i < num_var; ++i) {
+      res[i] = b[i];
+      for (size_t j = 0; j < dim; ++j) {
+        res[i] += A[i][j] * x[j];
+      }
+    }
+    return res;
+  };
+
+  // Create data data structures
+  std::array<std::unique_ptr<double[]>, dim> X_data;
+  Index<dim> num_x_points;
+
+  // Initialize discrete grid for the tabulated interpolation routine
+
+  size_t num_total = 1;
+  for (size_t n = 0; n < dim; ++n) {
+    num_x_points[n] = dist_int(gen);
+    num_total *= num_x_points[n];
+
+    X_data[n] = std::move(std::make_unique<double[]>(num_x_points[n]));
+
+    for (size_t m = 0; m < num_x_points[n]; m++)
+      X_data[n][m] = -1. + m * 2. / static_cast<double>(num_x_points[n] - 1);
+  }
+
+  // Correct for num_vars
+  num_total *= num_var;
+
+  // Allocate storage for main table
+  auto Y = std::make_unique<double[]>(num_total);
+
+  std::array<gsl::span<double const>, dim> Xarray;
+
+
+  // We are manually unscrambeling the compressed index here
+
+  for(size_t ijk=0; ijk < num_total/num_var; ++ijk){
```

Comment:
Is this basically the opposite of `collapsed_index` in `Index.hpp`? That is, given some extents and a 1d index, get the 3d index?

Reply:
Exactly.

Reply:
I cleaned up the comments to make this more clear